### PR TITLE
Add Fallout 4 support (#37)

### LINKED ISSUE #37: Manage Fallout 4
## Outline

The desired outcome of this issue is that Fallout 4:

- is automatically detected;
- can be managed by Nexus Mods App;
- game version is detected;

...for all applicable game stores.

## Supporting Information

- Automatic Game Detection
  - [ ] Steam
  - [ ] Windows Game Store
- Game Version Detection
  - [ ] Steam
  - [ ] Windows Game Store
- Game Managed
  - [ ] Steam
  - [ ] Windows Game Store

### PR DESCRIPTION
**Pull request by JonathanFeenstra**
Fixes #37.

Unfortunately, I couldn't get the unit test `CanInstallAndApplyMostPopularMods` to pass (I'm getting an `UnauthorizedAccessException` after the last line that I left uncommented), so I made this a draft PR for now.

### CODE REVIEW

**Comment by Sewer56** [comment]
(I'll review this, then have someone else doublecheck after. Hence I tagged this up earlier. Just getting some of my own PR work done first.)

**Review by Sewer56 — CHANGES_REQUESTED** [review]
I reviewed and made the following changes to bring it closer to a mergeable state:

- Added Fallout 4 to MkDocs Documentation Sidebar (was missing)
- Fixed the new test CanInstallAndApplyMostPopularMods
  - It used parameters we don't have for CLI commands, maybe a leftover from old App version.
- Removed unused field in test.

The PR gets an 'OK' from me, however we can't merge it yet.

```
NexusMods.Games.BethesdaGameStudios.Exceptions.MissingMastersException: Cannot sort plugins because:

NexusMods.Games.BethesdaGameStudios.Exceptions.MissingMastersException
Cannot sort plugins because: 
	Plugin Unofficial Fallout 4 Patch.esp requires master Fallout4.esm, but it is not present in the load order.
	Plugin Unofficial Fallout 4 Patch.esp requires master DLCRobot.esm, but it is not present in the load order.
	Plugin Unofficial Fallout 4 Patch.esp requires master DLCworkshop01.esm, but it is not present in the load order.
	Plugin Unofficial Fallout 4 Patch.esp requires master DLCCoast.esm, but it is not present in the load order.
	Plugin Unofficial Fallout 4 Patch.esp requires master DLCworkshop02.esm, but it is not present in the load order.
	Plugin Unofficial Fallout 4 Patch.esp requires master DLCworkshop03.esm, but it is not present in the load order.
	Plugin Unofficial Fallout 4 Patch.esp requires master DLCNukaWorld.esm, but it is not present in the load order.
```

I think we need to give the vanilla `plugins.txt` to `AddUniversalGameLocator`. Possibly do something else too. I'm pulling the GOG version of the game via the Nexus account, will see what's up soon.

For someone else interested in debugging, remember to set `NEXUS_API_KEY` environment variable.

**Comment by erri120** [comment]
I'm not sure if we want basic Fallout 4 support at this point. We're currently on milestone 0.4 which is all about Stardew Valley. Bethesda games will come in a later milestone.

**Comment by Sewer56** [comment]
Okay, some notes.
I tested with the GOG release. I don't have the Steam release available on me.

- Game appears to list 'plugins' in `%localappdata%/Fallout4/DLCList.txt`
- No `plugins.txt` is created during gameplay.

`DLCList.txt` is only created upon first run. i.e. if the user installs the game, but doesn't run it, it won't exist.
This should be mentioned in `0008-Fallout4.md`.

-------

Anyway, some caveats:

- We may want to generate a `plugins.txt` ourselves (or whatever would normally be in it).
    - Based on initial game folder contents.
    - Assume user never ran the game, so no reading `DLCList.txt`.
    - AFAIK, Steam version doesn't ship with the DLC, GOTY version everywhere else does.
    
This at the very least, would be required for 'basic support', so the end user can use the App at all.

Down the road:

- Error handling needed for when Mods require dependencies ('masters') not present in user's game copy.
    - e.g. Someone without DLC can't install `Unofficial Fallout 4 Patch`.  
    - Something made by design that's user friendly. 
    
In any case, the focus for next release is Stardew, so chances are none of us will pick this up till that point. 

I did review all of the code up till now, and made fixes (where needed) to match the original intent, but as long as we can't detect the core/standard 'plugins' present in game folder (i.e. generate a `plugins.txt` from that), the end user can't really do anything at all with NMA. 

I believe the method `BethesdaLoadoutSynchronizer.FixupLoadout` already exists for this, just need to patch that up to include an initial 'plugins.txt' file we can utilize.

**Comment by JonathanFeenstra** [comment]
I corrected the info about `plugins.txt` in the docs, I forgot to update that part after copying it from the Skyrim docs.

It's okay if this PR can't be merged until a later milestone, I just wanted to share my local changes from when I started working on this back in August. Unfortunately, I'm still getting the `UnauthorizedAccessException` when running the test, so I can't verify whether it's working.

**Comment by Sewer56** [comment]
I could probably reboot into Windows and test it there. On Linux at least, the test ran fine, other than the fact some logic is needed to construct a plugins file from initial game state.

**Comment by Sewer56** [comment]
Alright, I can confirm it

```
  Message: 
System.UnauthorizedAccessException : Access to the path 'CustomControlMap.txt' is denied.

  Stack Trace: 
FileSystem.RemoveDirectoryRecursive(String fullPath, WIN32_FIND_DATA& findData, Boolean topLevel)
FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
FileSystem.InternalDeleteDirectory(AbsolutePath path, Boolean recursive)
FileSystem.InternalDeleteDirectory(AbsolutePath path, Boolean recursive)
BaseFileSystem.DeleteDirectory(AbsolutePath path, Boolean recursive)
TemporaryPath.Dispose_Impl()
TemporaryPath.DisposeAsync()
FileOriginRegistry.RegisterDownload(AbsolutePath path, AArchiveMetaData metaData, CancellationToken token) line 80
NxmDownloadProtocolHandler.Handle(Uri url, LoadoutMarker loadout, String modName, CancellationToken token) line 58
NxmDownloadProtocolHandler.Handle(Uri url, LoadoutMarker loadout, String modName, CancellationToken token) line 64
<7 more frames...>
<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
InvocationPipeline.<InvokeAsync>g__FullInvocationChainAsync|2_0(InvocationContext context)
ParseResultExtensions.InvokeAsync(ParseResult parseResult, IConsole console)
ParserExtensions.InvokeAsync(Parser parser, String[] args, IConsole console)
CommandLineConfigurator.RunAsync(String[] args, IRenderer renderer, CancellationToken token)
AVerbTest.Run(String[] args) line 32
Fallout4Tests.CanInstallAndApplyMostPopularMods() line 38
```

Can confirm, the stack trace is this.
Entered by

```

File: docs/games/0008-Fallout4.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  
+Mods can ship as 'loose files' or '.esp+.ba2' pairs.
+
+## Additional Considerations for Manager
+
+### DLCList.txt
+
+!!! info "Found in: `%LOCALAPPDATA%\Fallout4` after running the game for the first time."
+
+The `DLCList.txt` file in Fallout 4 lists plugin files with the ".esp" and ".esm" extensions that are loaded after the game's own.
```
Comment (MattSturgeon) [inline docs/games/0008-Fallout4.md:32]: I believe `plugins.txt` also works. Or at least, the file created by Vortex seems to work fine with both the Steam and GOG versions in my experience.

Not sure if there's any preference or difference between `DLCList.txt` and `plugins.txt`, but the latter should probably get a mention in the docs?

**Comment by erri120** [comment]
I'm in favor of closing this PR for now. Fallout 4 support will be added later, currently we're still working on some core designs before continuing with game support.

**Comment by HerrTiSo** [comment]
Any rough ETA for Fallout 4 support, however? Especially considering the current hype around the IP thanks to the Amazon Prime show. Thanks for your hard work!

**Comment by erri120** [comment]
I'm closing this PR because we're going to add Fallout 4 support later. We need to build out core features first before adding support for Bethesda games.

**Pull request closed without merging**